data/transactions/logic/opcodes.go

```diff
@@ -61,38 +63,112 @@ const createdResourcesVersion = 6
 // experimental-
 const fidoVersion = 7 // base64, json, secp256r1
 
+type linearCost struct {
+	baseCost  int
+	chunkCost int
+	chunkSize int
+}
+
+func (lc *linearCost) compute(stack []stackValue) int {
+	cost := lc.baseCost
+	if lc.chunkCost != 0 && lc.chunkSize != 0 {
+		cost += lc.chunkCost * (len(stack[len(stack)-1].Bytes) + lc.chunkSize - 1) / lc.chunkSize
+	}
+	return cost
+}
+
+func (lc *linearCost) docCost() string {
+	if *lc == (linearCost{}) {
+		return ""
+	}
+	if lc.chunkCost == 0 {
+		return strconv.Itoa(lc.baseCost)
+	}
+	if lc.chunkSize == 1 {
+		return fmt.Sprintf("%d + %d per byte", lc.baseCost, lc.chunkCost)
+	}
+	return fmt.Sprintf("%d + %d per %d bytes", lc.baseCost, lc.chunkCost, lc.chunkSize)
+}
+
 // opDetails records details such as non-standard costs, immediate
 // arguments, or dynamic layout controlled by a check function.
 type opDetails struct {
-	Cost       int
-	costFunc   opCostFunc
+	FullCost   linearCost // if non-zero, the cost of the opcode, no fields matter
 	Size       int
 	checkFunc  opCheckFunc
 	Immediates []immediate
 	typeFunc   opTypeFunc
 }
 
-type opCostFunc func(program []byte, pc int) int
+func (d *opDetails) docCost() string {
+	cost := d.FullCost.docCost()
+	if cost != "" {
+		return cost
+	}
+	found := false
+	for _, imm := range d.Immediates {
+		if imm.fieldCosts != nil {
+			if found {
+				panic("two cost dependent fields")
+			}
+			found = true
+			group := imm.Group
+			for _, name := range group.Names {
+				fs, ok := group.SpecByName(name)
+				if !ok {
+					continue
+				}
+				cost += fmt.Sprintf(" %s=%d", name, imm.fieldCosts[fs.Field()])
+			}
+		}
+	}
+	return cost
+}
+
+// Cost computes the cost of the opcode, given details about how it is used,
+// both static (the program, which can be used to find the immediate values
+// supplied), and dynamic (the stack, which can be used to find the run-time
+// arguments supplied). Cost is used at run-time. docCost returns similar
+// information in human-reable form.
+func (d *opDetails) Cost(program []byte, pc int, stack []stackValue) int {
+	cost := d.FullCost.compute(stack)
```

Comment:
@jannotti Sanity checking my understanding to see if I've missed intuition:
* `Cost` treats base + stack input size vs field costs as mutually exclusive.
* Is the following statement accurate?
  * The mutual exclusivity simplifies cost management and we _may_ later revisit the invariant based on opcode modifications/additions.

Reply:
Yes, we will need to revisit this if we want immediates to have linear costs.  For example, if we moved to a single `hash` opcode that took the type of hash to do as an immediate. We'd just need to make `fieldCosts` an array of `linearCost` instead of simple `int` (and do the math, of course).

Reply:
@jannotti Thanks for the confirmation here - resolving.